File: Excel/684 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/684 Align Name and Data.xlsx"
input = read_excel(path, range = "A2:A14")
test  = read_excel(path, range = "C2:D6") %>% 
  replace_na(list(Amounts = " "))                 


result = input %>% 
  mutate(Name = ifelse(str_detect(Data, "\\d"), NA, Data)) %>%
  fill(Name) %>%
  mutate(Data = ifelse(Data == "Robert", " ", Data)) %>%
  filter(Data != Name) %>%
  summarize(Amounts  = paste0(Data, collapse = ", "), .by = Name)

all.equal(result, test)
#> [1] TRUE

File: Excel/684 Challenge.py
import pandas as pd
import numpy as np

path = "684 Align Name and Data.xlsx"

input_data = pd.read_excel(path, usecols="A", skiprows=1, nrows=13, names=["Data"])
test = pd.read_excel(path, usecols="C:D", skiprows=1, nrows=4).fillna({"Amounts": " "}).sort_values(by="Name").reset_index(drop=True)

input_data["Name"] = input_data["Data"].where(input_data["Data"].str.isalpha()).ffill()
input_data["Data"] = np.where(input_data["Data"] == "Robert", " ", input_data["Data"])
filtered_data = input_data[input_data["Data"] != input_data["Name"]]

grouped_data = filtered_data.groupby("Name")["Data"].apply(lambda x: ", ".join(map(str, x))).sort_index()
grouped_data = grouped_data.reset_index(name="Data")
grouped_data["Amounts"] = test["Amounts"].values
grouped_data = grouped_data.drop(columns=["Data"])


print(grouped_data.equals(test)) # True